experimental/yarpl/include/yarpl/Flowable.h

```diff
@@ -2,48 +2,87 @@
 
 #pragma once
 
-#include <functional>
 #include <memory>
-#include "yarpl/Flowable_Subscriber.h"
+#include <type_traits>
+#include <utility>
+#include "yarpl/flowable/sources/Flowable_RangeSubscription.h"
+
+#include "reactivestreams/ReactiveStreams.h"
+#include "yarpl/utils/type_traits.h"
 
 namespace yarpl {
 namespace flowable {
 
+// using reactivestream_yarpl to not conflict with the other reactivestreams
+
 template <typename T>
-class Flowable {
+class Flowable : public reactivestreams_yarpl::Publisher<T>,
+                 public std::enable_shared_from_this<Flowable<T>> {
+  using Subscriber = reactivestreams_yarpl::Subscriber<T>;
+  using Subscription = reactivestreams_yarpl::Subscription;
+
  public:
-  explicit Flowable(
-      std::function<void(std::unique_ptr<Subscriber<T>>)> onSubscribe);
+  template <
+      typename F,
+      typename = typename std::enable_if<
+          std::is_callable<F(std::unique_ptr<Subscriber>), void>::value>::type>
+  static std::unique_ptr<Flowable> create(F&& onSubscribeFunc) {
+    return std::make_unique<FlowableOnSubscribeFunc<F>>(
+        std::forward<F>(onSubscribeFunc));
+  }
 
-  static std::unique_ptr<Flowable<T>> create(
-      std::function<void(std::unique_ptr<Subscriber<T>>)> onSubscribe);
+  template <
+      typename F,
+      typename = typename std::enable_if<
+          std::is_callable<F(T), typename std::result_of<F(T)>::type>::value>::
+          type>
+  auto map(F&& function);
 
-  void subscribe(std::unique_ptr<Subscriber<T>>);
+  static std::unique_ptr<Flowable<long>> range(long start, long count) {
+    return create([start, count](auto subscriber) {
+      auto s_ = new yarpl::flowable::sources::RangeSubscription(
+          start, count, std::move(subscriber));
+      s_->start();
+    });
+  }
+
+ protected:
+  Flowable() = default;
+  Flowable(Flowable&&) = delete;
+  Flowable(const Flowable&) = delete;
+  Flowable& operator=(Flowable&&) = delete;
+  Flowable& operator=(const Flowable&) = delete;
 
  private:
-  std::function<void(std::unique_ptr<Subscriber<T>>)> onSubscribe_;
+  template <typename Function>
+  class FlowableOnSubscribeFunc : public Flowable {
+   public:
+    FlowableOnSubscribeFunc(Function&& function)
+        : function_(Function(std::forward<Function>(function))) {}
```

Comment:
If I understand correctly, because this is not a universal reference, despite looking like one, because the template is defined at the class level?